File: driver/pacific/shared/src/common/bit_utils.cpp
// BEGIN_LEGAL
//
// Copyright (c) 2017-current, Cisco Systems, Inc. ("Cisco"). All Rights Reserved.
//
// This file and all technical concepts, proprietary knowledge, algorithms and
// intellectual property rights it contains (collectively the "Confidential Information"),
// are the sole propriety information of Cisco and shall remain at Cisco's ownership.
// You shall not disclose the Confidential Information to any third party and you
// shall use it solely in connection with operating and/or maintaining of Cisco's
// products and pursuant to the terms and conditions of the license agreement you
// entered into with Cisco.
//
// THE SOURCE CODE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED.
// IN NO EVENT SHALL CISCO BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
// AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH
// THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
// END_LEGAL

#include "common/bit_utils.h"
#include <algorithm>
#include <stddef.h>

namespace silicon_one
{

namespace bit_utils
{

void
set_bits(uint64_t* target, size_t msb, size_t lsb, const uint64_t* val)
{
    size_t bytes_offset = lsb / BITS_IN_UINT64;
    target += bytes_offset;

    size_t bits_to_write = msb - lsb + 1, offset = lsb % BITS_IN_UINT64, val_offset = 0;

    while (bits_to_write > 0) {
        size_t iteration_bits_to_write = std::min({bits_to_write, BITS_IN_UINT64 - offset, BITS_IN_UINT64 - val_offset});
        uint64_t tmp = get_bits(*val, val_offset + iteration_bits_to_write - 1, val_offset);
        tmp = set_bits(*target, offset + iteration_bits_to_write - 1, offset, tmp);
        *target = tmp;
        bits_to_write -= iteration_bits_to_write;
        offset += iteration_bits_to_write;
        if (offset == BITS_IN_UINT64) {
            offset = 0;
            target++;
        }
        val_offset += iteration_bits_to_write;
        if (val_offset == BITS_IN_UINT64) {
            val_offset = 0;
            val++;
        }
    }
}

void
get_bits(const uint64_t* source, size_t msb, size_t lsb, uint64_t* out_val)
{
    size_t bytes_offset = lsb / BITS_IN_UINT64;
    source += bytes_offset;
    size_t out_val_offset = 0, bits_to_get = msb - lsb + 1, offset = lsb % BITS_IN_UINT64;

    while (bits_to_get > 0) {
        size_t iteration_bits_to_read = std::min({bits_to_get, BITS_IN_UINT64 - offset, BITS_IN_UINT64 - out_val_offset});
        uint64_t tmp = get_bits(*source, offset + iteration_bits_to_read - 1, offset);
        tmp = set_bits(*out_val, out_val_offset + iteration_bits_to_read - 1, out_val_offset, tmp);
        *out_val = tmp;
        bits_to_get -= iteration_bits_to_read;
        offset += iteration_bits_to_read;
        if (offset == BITS_IN_UINT64) {
            offset = 0;
            source++;
        }
        out_val_offset += iteration_bits_to_read;
        if (out_val_offset == BITS_IN_UINT64) {
            out_val_offset = 0;
            out_val++;
        }
    }
}

} // namespace bit_utils

} // namespace silicon_one

File: driver/gibraltar/shared/src/common/bit_utils.py
# BEGIN_LEGAL
#
# Copyright (c) 2020-current, Cisco Systems, Inc. ("Cisco"). All Rights Reserved.
#
# This file and all technical concepts, proprietary knowledge, algorithms and
# intellectual property rights it contains (collectively the "Confidential Information"),
# are the sole propriety information of Cisco and shall remain at Cisco's ownership.
# You shall not disclose the Confidential Information to any third party and you
# shall use it solely in connection with operating and/or maintaining of Cisco's
# products and pursuant to the terms and conditions of the license agreement you
# entered into with Cisco.
#
# THE SOURCE CODE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED.
# IN NO EVENT SHALL CISCO BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
# AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH
# THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# END_LEGAL


def get_all_1(num_of_bits):
    n = ((1 << num_of_bits) - 1)
    return n


def get_size_in_bits(n):
    if n > 0:
        size = 0
        my_n = n
        while (my_n > 0):
            size += 1
            my_n >>= 1
    else:
        size = 1
    return size


def get_bit(n, bit_index):
    bit = (n >> bit_index) & 0x1
    return bit


def set_bit(n, bit_index, data_to_set):
    new_n = set_bits(n, bit_index, bit_index, data_to_set)
    return new_n


def get_bits(n, msb, lsb):
    bit = (n >> lsb) & get_all_1(msb - lsb + 1)
    return bit


def set_bits(n, msb, lsb, data_to_set):
    data_to_set = data_to_set & get_all_1(msb - lsb + 1)
    new_n = n - (get_bits(n, msb, lsb) << lsb) + (data_to_set << lsb)
    return new_n


def bit_invert(n, size):
    mask = get_all_1(size)
    inverted_n = ~n & mask
    return inverted_n


def bit_reversal(x_in, size):
    x_out = 0
    for i in range(0, size):
        x_out = set_bit(x_out, size - 1 - i, get_bit(x_in, i))
    return x_out


def get_asserted_indexes(vector):
    n = get_size_in_bits(vector)
    for i in range(n):
        if (get_bits(vector, i, i) == 1):
            print("Asserted index = %0d" % (i))
